### ISSUE TITLE
Form for "Manage file display" attempts to load all field formatter forms (elements)

### ISSUE THREAD

**Issue by indigoxela** [backdrop/backdrop-issues#6578]
## Description of the bug

Possibly there's some weird reason for that, but it just doesn't seem right to me.

When configuring the "file display" (note: NOT the fields of the fieldable entity, but the file itself), all field formatters for file AND image show up as option in the radios. Regardless of the File type, or which mime types this File type's responsible for. But there's no field in this context. Which means, many (most) field formatter settings forms have to do some tricks to prevent PHP warnings because of missing field info.

## Steps To Reproduce

1. Install some modules that ship with field formatters
2. Go to /admin/structure/file-types/manage/document/file-display (which is the file display setting for document File types
3. Scratch your head

![document-file-display-options-form](https://github.com/backdrop/backdrop-issues/assets/842066/65f9fcf8-76f1-43bd-abbb-b383b0d478e2)

## Actual behavior

Field formatters show up as options, that make no sense at all.

Additionally, if these formatters don't catch that already, some PHP warnings are thrown because $field is NULL:

```
function hook_field_formatter_settings_form($field, $instance, $view_mode, $form, &$form_state) {
```

## Expected behavior

No completely unrelated options, I guess.

## Additional information

Add any other information that could help, such as:
- Backdrop CMS version: 1.28.0

**Comment by indigoxela** [comment]
Some more recherche:

Function file_info_formatter_types() loads all file_formatter_info() hook implementations via file_file_formatter_info().

So _image field formatters_ will always show up as options. Although that makes no sense in most cases.

**Comment by klonos** [comment]
> Install some modules that ship with field formatters

Interesting problem indeed @indigoxela ...can you please list some of these modules, so that we can easily reproduce the problem?

**Comment by indigoxela** [comment]
Try:

- Photoswipe
- Gallery tiler
- Colorbox
- Background Images Formatter

Basically any module that defines a formatter for image.

**Comment by klonos** [comment]
@indigoxela I believe that the problem is with the implementations of `hook_field_formatter_info()` in these contrib modules - not with core. They should be specifying the following (for images):
```php

      'file formatter' => array(
        'mime types' => array('image/*'),
      ),
```

Perhaps we could be deriving that from whatever is specified in the `'field types'` key of that hook implementation, but not 100% sure if that's the right way to go about it 🤔 ...besides, we only have two file-related field types in core:
- `image` ...which by pure luck would work for this use case specifically (since you are reporting this for image-related modules)
- `file` ...which is very broad and encapsulates all non-image field types

Unfortunately, we don't have any deeper separation of file-related field types - it's either images, or generic, non-image files 🤷🏼 

...it's the same in Drupal by the way:
![image](https://github.com/backdrop/backdrop-issues/assets/2423362/5c71fbf0-d303-41d3-9e1e-ea4bf432e136)

So perhaps that's what we'd need to address in core(?) ...in the meantime, the burden for fixing this should fall on the respective contrib modules (to "properly" implement their `hook_field_formatter_info()` and specify an appropriate `'file formatter'`).

**Comment by klonos** [comment]
...I actually have an idea that might work as a temporary measure, but it will be very limited in scope (only solve the problem for modules that are image-related formatters). But hey, something is better than nothing, right?

I'd like to give it a go.

**Comment by indigoxela** [comment]
I belief, this should be fixed in file_file_formatter_info https://github.com/backdrop/backdrop/blob/1.x/core/modules/file/file.module#L598

I currently "blindly" collects `array('file', 'image')` - at this point there isn't even a check, whether the image module in turned on, BTW.

Forcing modules to also check the mime type (for fields) seems an odd approach to me. It's not even documented that way. https://docs.backdropcms.org/api/backdrop/core%21modules%21field%21field.api.php/function/hook_field_formatter_info/1

A field formatter defines a ... **formatter for a field type** not a mime type. Core handles that wrong in function file_display_form() - it abuses "field formatters" as "file renderer".
So IMO leaving the responsibility to fix that to the modules defining field formatters actually makes things worse.

**Comment by kiamlaluno** [comment]
> it abuses "field formatters" as "file renderer"

I also think this is the issue: trying to use a field formatter where there aren't fields.

Field formatters and file formatters are two different concepts that should be kept separated. 

Furthermore:

- A file formatter needs to define the view callback in [`hook_file_formatter_info()`](https://docs.backdropcms.org/api/backdrop/core%21modules%21file%21file.api.php/function/hook_file_formatter_info/1), while a field formatter does not need to do that in [`hook_field_formatter_info()`](https://docs.backdropcms.org/api/backdrop/core%21modules%21field%21field.api.php/function/hook_field_formatter_info/1)
- The parameters for [`hook_field_formatter_view()`](https://docs.backdropcms.org/api/backdrop/core%21modules%21field%21field.api.php/function/hook_field_formatter_view/1) (the view callback for field formatters) and the parameters defined for [`hook_file_formatter_FORMATTER_view()`](https://docs.backdropcms.org/api/backdrop/core%21modules%21file%21file.api.php/function/hook_file_formatter_FORMATTER_view/1) (the view callback for file formatters) are totally different

**Comment by kiamlaluno** [comment]
[`file_file_formatter_file_field_view()`](https://docs.backdropcms.org/api/backdrop/core%21modules%21file%21file.module/function/file_file_formatter_file_field_view/1) has a comment with the following sentences.

> Note that we are reusing field formatter functions, but we are not displaying a Field API field, so we set `$field` and `$instance` accordingly, and do not invoke `hook_field_prepare_view()`. This assumes that the formatter functions do not rely on `$field` or `$instance`. A module that implements formatter functions that rely on `$field` or `$instance` (and therefore, can only be used for real fields) can prevent this formatter from being used on the pseudo-field by removing it within `hook_file_formatter_info_alter()`.

That assumes a field formatter does not need to get information about the field it should format, and that invoking `hook_field_prepare_view()` is not necessary.
IMO, that are wrong assumptions, which make me think `hook_field_formatter_info()` is used for the wrong purpose.

**Comment by kiamlaluno** [comment]
As a side note, all the options shown for _Display type_ in the first screenshot are field formatters. *Generic file*, *Table of files*, *URL to file*, and *Download link* are formatters returned from [`file_field_formatter_info()`](https://docs.backdropcms.org/api/backdrop/core%21modules%21file%21file.field.inc/function/file_field_formatter_info/1).

**Comment by indigoxela** [comment]
> all the options shown for Display type in the first screenshot are field formatters

Yes, this "hybrid character" of field formatters makes it all so tricky. Contribs can only "opt out" here with some tricks, but developers need to figure out first, that field formatters are used for something completely different.
So either there's some crucial documentation missing (currently it's rather misleading), or the whole concept of this "Manage file display" is fishy. :stuck_out_tongue_winking_eye:

**Comment by indigoxela** [comment]
We should - to the very least - document somewhere, with some emphasis of the importance, that modules providing **field formatters for image fields** are required to also implement hook_file_formatter_info_alter() to prevent core to abuse their field formatter as a file formatter.

Creating field formatters is a challenge on its own, as lots of hooks are necessary for that, already. We shouldn't let developers blunder into that trap with file display, which isn't documented anywhere with field formatter hooks.
I created several field formatters, but until now wasn't aware of this problem.

Not that I think, documentation only is the right solution, but I don't see a way to fix that hook abuse in core without major changes in the whole logic, with possibly breaking changes.

**Comment by kiamlaluno** [comment]
The `File` class implements an entity. Instead of using a pseudo-field which is then rendered using field formatters (which expect entity fields), `File` should have a field for the file content, which is then rendered using a field formatter, which would then get a real entity field.  
It should not be different from the body field used from nodes.

**Comment by indigoxela** [comment]
Let's fix this bug! Every time I create a field formatter (for any type of field), I have to remember to "opt out" of file display. Let's make it "opt in" instead. PR available for testing and review: https://github.com/backdrop/backdrop/pull/5038

**Comment by laryn** [comment]
Thank you @indigoxela, @avpaderno, @klonos, and @bugfolder! Merged into `1.x1 and `1.31.x`.

**Comment by argiepiano** [comment]
This introduces a regression that has been reported [here](https://github.com/backdrop/backdrop-issues/issues/6969). 

The problem is fairly simple to spot: custom file field formatters provided by custom modules like Colorbox are removed by this PR, since it only allows formatters provided by the module `file` to show up.   

EDIT: a bit more explanation.  Before this PR, `file_file_formatter_info()` reused  all  file and image _Field API field_ formatters, and created new _file entity formatters_ based on those Field API field formatters. This included file and image field formatters defined by custom modules. After this PR, only those field formatters provided by the core `file` module were allowed to be reused. So, if you had views of file entities that originally used a refactored custom field formatter (e.g. Colorbox's image field formatter),  those are gone after this PR. 

Ideally, we wouldn't need to reuse _field_ formatters as _file entity formatters_. Each module like Colorbox would provide a file formatter, in addition to a field formatter. One of the gray areas here is that file entity bundles ("file entity types" such as Document, Image and Video)  are not necessarily tied to a specific file mime type - you can edit the mime types "accepted" by a specific file entity bundle. So, you can't simply say that the "Document" file entity bundle will always only accept "non images" (or files that the browser can't render). And you can technically create a new file entity bundle that accepts both some file mime types that can be rendered by the browser, and others that can't. 

So, with the current architecture, you can't really limit a file formatter to a specific file entity bundle - you are stuck with showing ALL file formatters for ALL file entity bundles, unless we somehow change the architecture.

**Comment by herbdool** [comment]
We'll need a follow-up issue for folks to find an alternative PR. For now, we'll probably revert the fix introduced here.
